kclvm/tools/src/LSP/src/rename.rs

```diff
@@ -0,0 +1,302 @@
+/// pkg_root: the absolute file path to the root package
+/// file_paths: list of files in which symbols can be renamed
+/// symbol_path: path to the symbol. The symbol path should be in the format of: `pkg.sub_pkg:name.sub_name`
+/// new_name: the new name of the symbol
+pub fn rename_symbol(
+    pkg_root: &str,
+    file_paths: &[String],
+    symbol_path: &str,
+    new_name: String,
+) -> Result<HashMap<Url, Vec<TextEdit>>, String> {
+    // 1. from symbol path to the symbol
+    match parse_symbol_selector_spec(pkg_root, symbol_path) {
+        Ok(symbol_spec) => {
+            if let Some((name, def)) = select_symbol(&symbol_spec) {
+                if def.is_none() {
+                    return Result::Err(format!(
+                        "can not find definition for symbol {}",
+                        symbol_path
+                    ));
+                }
+                match def.unwrap().get_kind() {
+                    SymbolKind::Unresolved => {
+                        return Result::Err(format!(
+                            "can not resolve target symbol {}",
+                            symbol_path
+                        ));
+                    }
+                    _ => {
+                        // 3. build word index on file_paths, find refs within file_paths scope
+                        if let Ok(word_index) = build_word_index_for_file_paths(file_paths) {
+                            if let Some(locations) = word_index.get(&name) {
+                                // 4. filter out the matched refs
+                                // 4.1 collect matched words(names) and remove Duplicates of the file paths
+                                let file_map = locations.iter().fold(
+                                    HashMap::<Url, Vec<&Location>>::new(),
+                                    |mut acc, loc| {
+                                        acc.entry(loc.uri.clone()).or_insert(Vec::new()).push(loc);
+                                        acc
+                                    },
+                                );
+                                let refs = file_map
+                                    .iter()
+                                    .flat_map(|(_, locs)| locs.iter())
+                                    .filter(|&&loc| {
+                                        // 4.2 filter out the words and remain those whose definition is the target def
+                                        let p = loc.uri.path();
+                                        if let Ok((_, _, _, gs)) = parse_param_and_compile(
+                                            Param {
+                                                file: p.to_string(),
+                                            },
+                                            None,
+                                        ) {
+                                            let kcl_pos = kcl_pos(p, loc.range.start);
```

Comment:
Do not use references from GlobalState that has been destroied, please directly compare the ranges of two symbol.
You can do it like this:
```rust 
                                if let Ok((_, _, _, gs)) = parse_param_and_compile(
                                    Param {
                                        file: p.to_string(),
                                    },
                                    None,
                                ) {
                                    let kcl_pos = kcl_pos(p, loc.range.start);
                                    let symbol_ref = find_def_with_gs(&kcl_pos, &gs, true).unwrap();

                                    let symbol_def =
                                        gs.get_symbols().get_symbol(symbol_ref).unwrap();

                                    return symbol_def.get_range() == def_pos;
                                }
```

Reply:
fixed, ptal